Retrieve log urls

maruiz93 opened the pull request:
<!--  Thanks for sending a pull request!

Please read our contributor guidelines:
https://kubearchive.github.io/kubearchive/main/contributors/guide.html
-->

## Which Issue(s) this Pull Request Resolves
<!-- Please make sure to create an issue you can link to and mention it here
to automatically close it. If this PR covers part of the work, instead of
"Resolves #", use "Related to #"

Usage: `Resolves #<issue number>`, or `Resolves (paste link of issue)`.
-->

Resolves #539

## Release Note
<!--
If this change has user-visible impact, write a release note in the block
below. Include the string "action required" if additional action is required of
users switching to the new release, for example in case of a breaking change.

Write as if you are speaking to users, not other KubeArchive contributors. If this
change has no user-visible impact, leave the code block as is.
-->

```release-note
Add new endpoints for the retrieval of log URLs
```

## Notes for Reviewers
<!--
Leave notes for the reviewers if you want to point their attention to some
specific place.
-->
I've included a new file called `query.go` in the `database` package with helper functions to encapsulate the logic for parsing the sql row results into structs or primitive data.
All the queries (new and old) are using this new functions.

There is a new interface `DBParamParser` to be implemented in the different drivers for transforming arrays into the allowed parameters for parametrized statements. This way we allow having queries of the type `SELECT * FROM ... WHERE key IN ?` being `?` a set of elements of undefined length.

The endpoints return 404 when the resource doesn't exist in the database.

For testing, you can run the integration tests locally, check the namespace used for the generated data and then do:
```
kubectl apply -f test/users/
kubectl get -n kubearchive secrets kubearchive-api-server-tls -o jsonpath='{.data.ca\.crt}' | base64 -d > ca.cr
curl -v --cacert ca.crt -H "Authorization: Bearer $(kubectl create -n test token default)" \
 https://localhost:8081/apis/batch/v1/namespaces/<ns>/cronjobs/generate-log-1/log
```

Unit tests are included but no integration tests. I will open a new issue for addressing those.
<!--
Please label this pull request according to what type of issue you are addressing.
For reference on required PR/issue labels, read here:
https://kubearchive.github.io/kubearchive/main/contributors/release.html#_pull_request_labels

Add one of the following labels:
kind/bug
kind/documentation
kind/feature

Optionally add one or more of the following kinds if applicable:
kind/breaking
-->

skoved commented on pkg/database/database.go:
```diff
@@ -107,39 +114,71 @@ func (db *Database) QueryNamespacedResourceByName(ctx context.Context, kind, api
+		Uuid string `json:"uuid"`
+		Kind string `json:"kind"`
+	}
+	i := len(uuids)
+	for i > 0 {
+		parsedRows, err := newQueryPerformer[uuidKind](ctx, db.db).performQueryWithArray(db.info.GetOwnedResourcesSQL(), uuids)
+		if err != nil {
+			return nil, err
+		}
+		uuids = make([]string, 0)
+		for _, row := range parsedRows {
+			if row.Kind == "Pod" {
+				podUuids = append(podUuids, row.Uuid)
+			} else {
+				uuids = append(uuids, row.Uuid)
+			}
+		}
+		i = len(uuids)
+	}
```
I think the looping logic could be made more clear by having it all in the `for` line

```suggestion
	for i:= len(uuids); i > 0; i = len(uuids) {
		parsedRows, err := newQueryPerformer[uuidKind](ctx, db.db).performQueryWithArray(db.info.GetOwnedResourcesSQL(), uuids)
		if err != nil {
			return nil, err
		}
		uuids = make([]string, 0)
		for _, row := range parsedRows {
			if row.Kind == "Pod" {
				podUuids = append(podUuids, row.Uuid)
			} else {
				uuids = append(uuids, row.Uuid)
			}
		}
	}
```

maruiz93 replied:
great point, as I was thinking of a "while" translation I didn't thought of adding the initialize and the "post-loop" change on the `for`.

skoved commented on pkg/database/query.go:
would using [`sqlx`](https://pkg.go.dev/github.com/jmoiron/sqlx) make sense here? [`sqlx.SelectContext`](https://pkg.go.dev/github.com/jmoiron/sqlx#SelectContext) seems like it would do what `QueryPerformer` is does and we wouldn't have to maintain the logic ourselves

maruiz93 replied:
> On a more general note, why do we have to treat these queries different than any other queries?

Until now we needed to retrieve always the same stuff from all the selects, in other words, all the `SELECT` queries that we had the same fields: `uuid`, `created_at` and `data`

With the LOG_URL retrieval I am introducing new queries that retrieve different things:
1. A query that retrieves the `kind` and `uuid`, for the query that selects all the directly owned resources by a specific uuid
2. A query that retrieves the `url` field

So now we have 3 type of queries with regards to the selected fields, and that's why I introduced this new `QueryPerformer` parametrized struct.

If I instantiate a `QueryPerformer[string]` means that when I perform the query I would receive a slice of strings, if I instantiate a `QueryPerformer[int]` means that when I perform the query I would receive a slice of ints, and if I instantiate a `QueryPerformer[struct]` means that when I perform the query I would receive a slice of instances of that struct.

> would using [sqlx](https://pkg.go.dev/github.com/jmoiron/sqlx) make sense here? [sqlx.SelectContext](https://pkg.go.dev/github.com/jmoiron/sqlx#SelectContext) seems like it would do what QueryPerformer is does and we wouldn't have to maintain the logic ourselves

I think we could take a look at it but I have the feeling that it wouldn't be that simple, specially regarding passing arrays as arguments, but I would rather investigate that in a separate issue. BTW, note that I included a note on top on the "worst" function (the one that does the reflection) of the `query.go` file pointing to [an issue](https://github.com/golang/go/issues/61637) in sql/db library for implementing that logic that is already available in `sqlx` as you mention.

ggallen replied:
So are we planning to refactor all the other queries to use `QueryPerformer`?

skoved replied:
> I think we could take a look at it but I have the feeling that it wouldn't be that simple, specially regarding passing arrays as arguments, but I would rather investigate that in a separate issue

That's ok with me. My understanding is that the array that you pass in is supposed to be empty and gets populated with the data from each row of the query. I'll open an issue to investigate using sqlx

skoved reviewed (changes requested):
The lookups for log urls seem to me like the most complex set of queries KubeArchive has to do so far, so thanks for working on this! There's a couple things I noticed but this otherwise looks good so far

ggallen commented:
I started to review this and while it probably works it seems overly complicated.  The process of generating the log URLs appears to lend itself to recursion.  I think most of your changes can probably be replaced by something like this:
```
type uuidKind struct {
        Uuid string 
        Kind string
}

func (db *Database) QueryLogURLs(ctx context.Context, kind, apiVersion, namespace, name string) ([]string, error) {
        []string urls 
        string uuid = db.getUUID(ctx, kind, apiVersion, namespace, name)
        return getLogUrls(ctx, { Uuid: uuid, Kind: kind }, urls)
}

func (db *Database) getLogURLs(ctx context.Context, uuid uuidKind, urls []string) ([]string error) {
        if uuid.Kind == 'Pod' {
                return db.getLogURLs(uuid), nil
        } else {
                owners, _ := db.getOwnedResources(ctx, uuid.Uuid) // Return UUID/kind of resources that uuid owns.
                for owner := range owners { 
                        urls = append(urls, getLogURLs(ctx, owner, urls))
                }
                return urls, nil
        }
}
Note that this is pseudo code and has to be cleaned up, but I think it logically does what is needed.

rh-hemartin commented on cmd/api/routers/routers.go:
```diff
@@ -108,6 +108,27 @@ func (c *Controller) GetNamespacedResourceByName(context *gin.Context) {
 	context.JSON(http.StatusOK, resource)
 }
 
+func (c *Controller) GetLogsURLByResourceName(context *gin.Context) {
```
Just for consistency.

```suggestion
func (c *Controller) GetLogURLsByResourceName(context *gin.Context) {
```

maruiz93 replied:
done! and changed in both new entrypoint methods

rh-hemartin reviewed (approved):
Overall I think this is good. The `queryPerformer` is handy, but we are burying deeper the `db` connection. I'm willing to assume that tradeoff right now. We need to move forward, so I would revisit this after a while and see if and how we can simplify it.

maruiz93 commented:
> I started to review this and while it probably works it seems overly complicated. The process of generating the log URLs appears to lend itself to recursion. I think most of your changes can probably be replaced by something like this:
> 
> ```
> type uuidKind struct {
>         Uuid string 
>         Kind string
> }
> 
> func (db *Database) QueryLogURLs(ctx context.Context, kind, apiVersion, namespace, name string) ([]string, error) {
>         []string urls 
>         string uuid = db.getUUID(ctx, kind, apiVersion, namespace, name)
>         return getLogUrls(ctx, { Uuid: uuid, Kind: kind }, urls)
> }
> 
> func (db *Database) getLogURLs(ctx context.Context, uuid uuidKind, urls []string) ([]string, error) {
>         if uuid.Kind == 'Pod' {
>                 return db.getLogURLs(uuid), nil
>         } else {
>                 owners, _ := db.getOwnedResources(ctx, uuid.Uuid) // Return UUID/kind of resources that uuid owns.
>                 for owner := range owners { 
>                         urls = append(urls, getLogURLs(ctx, owner, urls))
>                 }
>                 return urls, nil
>         }
> }
> ```
> 
> Note that this is pseudo code and has to be cleaned up, but I think it logically does what is needed.

The problem I see with this is that you are assuming that there is only one jump between the resource given and the pods (e.g., a `TaskRun` and a `Pod` or a `ReplicaSet` and a `Pod` or a `Job` and a `Pod`) but you could have several levels (e.g. a `Deployment` owns a `ReplicaSet` that owns `Pod`s or a `CronJob` owns `Job`s that owns `Pod`s) and that's why we need some kind of recursion because the number of jumps could be n.

Also, note that with this proposal I am trying to reduce the number of queries to the minimum:

* If a pod is given just one query is being done
* If something different that a pod is given, then
    * A query is performed to get the uuid of the resource
    * n queries are performed to get the owned resources given a uuid, being n the number of jumps between the resource kind and the owned resources in the "ownership chain" (e.g, for a `Deployment` or a `CronJob` n=2, for a `ReplicaSet` or a `Job` n=1).
    * A query is performed to get the urls given the pod uuids

Considering this the total number of queries performed if the given resource is not a pod is: 2+n and I wouldn't expect to have n greater than 3.

ggallen commented:
> The problem I see with this is that you are assuming that there is only one jump between the resource given and the pods (e.g., a `TaskRun` and a `Pod` or a `ReplicaSet` and a `Pod` or a `Job` and a `Pod`) but you could have several levels (e.g. a `Deployment` owns a `ReplicaSet` that owns `Pod`s or a `CronJob` owns `Job`s that owns `Pod`s) and that's why we need some kind of recursion because the number of jumps could be n.

No, I'm not making that assumption at all. What I have coded handles all the cases you describe.  And I think it's significantly less code.

> If a pod is given just one query is being done

I think this is the only difference in the number of queries, and it's minimal.  It's only more in the case where you query logs for a `Pod` directly.  I would actually suggest "hiding" the `GetUUID` call behind the other calls. There's really no need to expose it to the users.

ggallen commented:
So here's a code version using recursion and your code. I haven't compiled it, but it should work and I think it's simpler than your loop.
```
type uuidKind struct {
	Uuid string `json:"uuid"`
	Kind string `json:"kind"`
}

func (db *Database) QueryLogURLs(ctx context.Context, kind, apiVersion, namespace, name string) ([]string, error) {
        []string urls
	err := db.db.QueryRowContext(ctx, db.info.GetUUIDSQL(), kind, apiVersion, namespace, name).Scan(&uuid)
	if err != nil {
		return nil, err
	}
        return queryLogURLs(ctx context.Context, {Uuid: uuid, Kind: kind }, urls)
}

func (db *Database) queryLogURLs(ctx context.Context, uuid uuidKind, []string urls) ([]string, error) {
	if uuid.Kind == 'Pod' {
	    return strQueryPerformer.performQuery(db.info.GetLogURLsSQL(), uuid.Uuid)
        } else {
		parsedRows, err := newQueryPerformer[uuidKind](ctx, db.db).performQuery(db.info.GetOwnedResourcesSQL(), uuid.Uuid)
		if err != nil {
			return nil, err
		}
                for _, row := range parsedRows {
			urls = append(urls, queryLogURLs(ctx, row, urls))
                }
                return urls, nil
        }
}
```
I note that you are correct that there will be more queries here as I am querying the log URLs for each `Pod` while you are getting them all in one call at the end.

But this recursion could easily be modified to simply get the `Pod` UUIDs instead and then make the same call at the end to get the log URLs.

maruiz93 commented:
Still working on the PR for the test stuff, but I addressed some suggestions and also dealt with the arrays for the parametrized queries, a thing that is tightly coupled to the driver used. I created a new interface called `ParamParser` that the different driver database should implement for specifying how to deal with parameters in the form of array (e.g. for the `select ... IN (<array>)`)

Please take a look :)

maruiz93 commented on pkg/database/query.go:
```diff
@@ -0,0 +1,138 @@
+
+type queryPerformer[T any] struct {
+	zeroVal T
+	ctx     context.Context
```
Very good point that I wasn't aware of. I've already changed it.

maruiz93 commented on pkg/database/query.go:
```diff
@@ -0,0 +1,138 @@
+func (q queryPerformer[T]) performQueryWithArray(query string, args []string) ([]T, error) {
+	// FIXME This need to be compatible with other drivers not just with postgres
+	return q.parseRows(q.db.QueryContext(q.ctx, query, pq.Array(args)))
+}
```
Done

maruiz93 commented on pkg/database/mariadb.go:
```diff
@@ -49,6 +49,23 @@ func (info MariaDBDatabaseInfo) GetNamespacedResourceByNameSQL() string {
 }
 
+func (info MariaDBDatabaseInfo) GetUUIDSQL() string {
+	return "SELECT uuid FROM resource WHERE kind=?AND api_version=? AND namespace=? AND name=?"
```
good catch! Done!

ggallen commented on pkg/database/database.go:
```diff
@@ -107,39 +127,81 @@ func (db *Database) QueryNamespacedResourceByName(ctx context.Context, kind, api
+		Kind string `json:"kind"`
+	}
+	// This loops for each jump existing in the ownerReference chain from the first resource until the pod
+	// The slice uuids is modified inside the loop
```
Loop counters should generally not be modified within the loop. This tells me something is wrong.
```suggestion
	pq := db.newParamQuery(db.info.GetLogURLsSQL())
	pq.addStringArrayParam(podUuids)
	return strQueryPerformer.performQuery(ctx, pq)
```

maruiz93 replied:
ok, so look at it in a different way, note that golang doesn't have the `while` reserve word, `while` is performed with the `for`. So, I am doing things `while` I still have `uuid`s of  resources different of a `Pod`, but I am not iterating over an array/slice. In other words, I am not modifying the counter of a loop, I am doing things while I still have uuids, and I am removing uuids in the process and adding new ones

ggallen replied:
You seem intent on ignoring my suggestion to use recursion to solve this problem, and I'm not sure why.

The fact that what you are trying to do isn't supported easily in golang indicates to me that you should consider other options.  What I am proposing works and I think is far more readable. I have been looking at this code for the past couple of days and I find it very confusing.  The fact that you have to explain in such detail also indicates that it's either overly complex or just unclear.

You seem to be focused on making the fewest number of calls possible, which is an admirable goal. But to me the tradeoff with the complexity and the lack of readability of the code makes it not worth it. Here are two examples I can cite:

1. The first call to special-case if the request is for a `Pod`.  Yeah, you can do that and you make one less call. But to me a more general solution that just always gets the uuid and then goes from there is easier to read and more concise. Everything is handled the same.  I'm not sure the tradeoff is worth it for the number of times logs will be requested in general versus the number of times logs are requested specifically for a `Pod`.
2. Batching uuids together for calls. I understand this a little more, but again I'm not sure it's worth the tradeoff. It makes no difference in your implementation, and it's clear why you are doing it. But I can tell you I didn't notice this batching the first N times I read the code, and as  a result I thought the code was broken, which again points to readability. Again, I question whether this is worth it, but recursion can also support doing it that way as well.

So I will meet you halfway and give you #2 but not number #1.  Here's yet another solution with that in mind:

```suggestion
type uuidKind struct {
	Uuid string `json:"uuid"`
	Kind string `json:"kind"`
}

func (db *Database) QueryLogURLs(ctx context.Context, kind, apiVersion, namespace, name string) ([]string, error) {
        []string PodUuids
        string owner
	err := db.db.QueryRowContext(ctx, db.info.GetUUIDSQL(), kind, apiVersion, namespace, name).Scan(&owner)
	if err != nil {
		return nil, err
	}
        podUuids, err := getPodUUIDs(ctx, [ owner ], podUuids)
	pq := db.newParamQuery(db.info.GetLogURLsSQL())
	pq.addStringArrayParam(podUuids)
	return strQueryPerformer.performQuery(ctx, pq)
}

func (db *Database) getPodUUIDs(ctx context.Context, []string owners, []string podUuids) ([]string, error) {
	if len(uuids) == 0 {
	    return podUuids, nil
        } else {
		pq := db.newParamQuery(db.info.GetOwnedResourcesSQL())
		pq.addStringArrayParam(owners)
		parsedRows, err := newQueryPerformer[uuidKind](db.db).performQuery(ctx, pq)
		if err != nil {
			return nil, err
		}

                var uuids string[]
                for _, row := range parsedRows {
			if row.Kind == 'Pod' {
				podUuids = append(podUuids, row.Uuid)
			} else {
				uuids = append(uuids, row.Uuid))
			}
                }
                return getPodUuids(ctx, uuids, podUuids)
        }
}
```
I don't like it as well as my previous solution, and I think it's uglier because it does the batching as you were. But I still think it's much more concise and readable that what you have.

> In other words, I am not modifying the counter of a loop, I am doing things while I still have uuids, and I am removing uuids in the process and adding new ones

I see your point here, but I think it's semantics. You aren't actually touching `i`. You are however assigning `i` a value rather than the normal increment.  It looks odd.

As I said, I don't like it, but this is even better IMO. You don't need `i` at all:
```
	for ; len(uuids) > 0;) {
```

maruiz93 replied:
> You seem intent on ignoring my suggestion to use recursion to solve this problem, and I'm not sure why.

I am sorry, it wasn't my intent to ignore you and I had a comment written these days that I thought I sent but I didn't press the "Comment" button in Github -.-'
I'll try to address the points that I did there and also what you mention in here

> The fact that what you are trying to do isn't supported easily in golang

I am not sure why you are saying this, it is supported, it's just the way golang handles the `while` of other languages.

> You seem to be focused on making the fewest number of calls possible

Indeed I am. 
Let's say you have a large complex PipelineRun, with lots of TaskRuns, let's say 15. 
If I batch the uuids, I will only do 2 queries to get the owned resources, if I go one by one as you suggested at some point, you will have 16 queries, one per PipelineRun and one per TaskRun to find its pods. 
I really believe we should prioritize optimization in the number of queries for this use case.

> So I will meet you halfway and give you `#2` but not `#1`

I could live with this, even though it makes more sense to me trying to optimize the number of queries in general.

> Regarding recursion

What I mentioned in my lost comment was that I could do recursion while keeping the logic that I have. My idea was actually pretty similar to your current suggestion but I was going to say let's do it in a separate MR.
One other thing that I was considering for the recursion is to make it a [tail-recursion](https://medium.com/@meeusdylan/tail-recursion-in-go-fb5cf69a0f26) as it is more efficient. 
Your previous suggestion weren't tail recursion functions but this one is, so it looks not good to me.

> Regarding readability

I guess that although this is kind of subjective, recursion is usually conceived as a more elegant and readable solution than iteration, and even though I chose iteration for my first approach, I don't have anything against recursion (I actually enjoy designing recursive algorithms while keeping tail-recursion in mind)

> Regarding the usage of `i` in the iteration definition

True, I actually like the `for ; len(uuids) > 0; {` that you suggest, it is cleaner.

> In summary

Now that you took the time to do a suggestion for the tail-recursive function that keeps batch queries, tomorrow I'll extract this part of the code to a recursive function following your suggestion.
I may want to try to keep the happy just-one-query path for the cases where pods are given, which is also very go-stylish as it means an early return in the function. We can discuss it tomorrow.

ggallen replied:
> True, I actually like the for ; len(uuids) > 0; { that you suggest, it is cleaner.

Actually, I just looked again and you can drop the semi-colons:

````
for len(uuids) > 0 {
        ...
}
```

ggallen replied:
If you *really* want to optimize, you could consider using a stored procedure here instead. Although stored procedures and individual calls perform about the same, the risk of a network issue during the calls is significantly higher for individual calls rather than the one stored procedure call.   :-)

I am *not* proposing this change, just pointing out that it might actually be more performant.

maruiz93 replied:
hmm, I am not against it if the performance isn't enough, but at first I believe that being able to receive intermediate results in the API may end up in delivering more information to the user of potential problems.

For now, I updated this logic from iteration to recursion as you suggested

maruiz93 commented on pkg/database/query.go:
```diff
@@ -0,0 +1,138 @@
+
+func (q queryPerformer[T]) performQuery(query string, args ...string) ([]T, error) {
+	return q.parseRows(q.db.QueryContext(q.ctx, query, toSliceOfAny(args)...))
+}
```
done

maruiz93 commented:
I still need to add new tests but I want to see the output of the CI

maruiz93 commented:
Added unit tests and the extra sar request for the `pod/logs` resource

rh-hemartin commented on cmd/api/auth/authz.go:
```diff
@@ -36,44 +37,59 @@ func RBACAuthorization(sari clientAuthzv1.SubjectAccessReviewInterface, cache *c
-			if allowed != true {
-				abort.Abort(c, "Unauthorized", http.StatusUnauthorized)
-				return
+		if strings.HasSuffix(c.Request.RequestURI, "log") {
```
```suggestion
		if strings.HasSuffix(c.Request.URL.Path, "/log") {
```

The `http.Request` docs recommend not using `.RequestURI` but `.URL` instead. Anyway, I also suggest you add a leading `/` so this don't trigger on resources that end with `log`. I don't think there will be many, but it can happen.

maruiz93 replied:
good point! Changes applied

rh-hemartin commented on pkg/database/query.go:
```diff
@@ -0,0 +1,154 @@
+
+type paramQuery struct {
+	query         string
+	arguments     []interface{}
```
You can use `any`, which sometimes is "plain english" than `interface{}`:


```suggestion
	arguments     []any
```

However it is not a big deal, at the end of the day both are used widely.

maruiz93 replied:
I changed my mind several times of using the alias over the native name or the other way around but at the end I have both and I see that are using the alias in other parts of the code so I'll change to the alias `any` for consistency.
